File: Computer Programming/C++/Recursion and Dynamic Programming/Unique Paths II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) 
    {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();    
        if(n==0 && m==0)
        {
            return 0;
        }
        if(n==1 && m==1)
        {
            if(obstacleGrid[0][0]==1)
            {
                return 0;
            }
            else
            {
                return 1;
            }
        }
        vector<vector<int>>vis(m,vector<int>(n,-1));

        return path(obstacleGrid,m-1,n-1,vis);
    }

    int path(vector<vector<int>>& obstacleGrid,int i,int j,vector<vector<int>>& vis)
    {
        if(i==0&&j==0)
        {
            return 1;
        }
        if(i<0 || j<0 || obstacleGrid[i][j]==1)
        {
            return 0;
        }
        if(vis[i][j]!=-1)
        {
            return vis[i][j];
        }
    int top=0;
    int left=0;
    if(i-1>=0)
    {
        if(obstacleGrid[i-1][j]==0)
        {
            top = path(obstacleGrid,i-1,j,vis);
        }
    }
    if(j-1>=0)
    {
        if(obstacleGrid[i][j-1]==0)
        {
            left = path(obstacleGrid,i,j-1,vis);
        }
    }

        vis[i][j] = top + left;
        return vis[i][j];
    }
};

File: Computer Programming/Python/Dynamic Programming/Unique Paths II.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        row = len(obstacleGrid)
        col = len(obstacleGrid[0])

        dp = []
        for i in range(0,row):
            curr = []
            for j in range(0,col):
                curr.append(-1)
            dp.append(curr)
        

        ans = self.function(dp,obstacleGrid,row-1,col-1)
        return ans

    def function(self,dp,obstacleGrid,row,col):
        if row == 0 and col == 0:
            return 1
        if row < 0 or col < 0:
            return 0
        if dp[row][col] != -1:
            return dp[row][col]
        
        if obstacleGrid[row][col] == 1:
            return 0

        left = self.function(dp,obstacleGrid,row-1,col)
        down = self.function(dp,obstacleGrid,row,col-1)

        dp[row][col] = left + down
        return dp[row][col]
